### PR TITLE
Some optimizations to the shrinker

### PR DESCRIPTION
**Pull request by DRMacIver**
This is some work extracted from my ongoing work on the new backend. It makes the shrinker a fair bit faster by some more judicious choices of shrinking - there is less wasted work and bigger shrinks are achieved faster.

It has three major features:

1. A bunch of tuning of the lexicographic minimizer. In particular this removes a bunch of bulk operations that always performed a lot of attempted shrinks and rarely worked, and generalises the ddmin stuff to a set of operations that seem to do a good enough job most of the time.
2. Better duplication of blocks in the main engine - a new pass which tries many clones at once on the off chance that they'll succeed.
3. Move the whole buffer lexicographic shrink later in the process so it's not doing a lot of duplicated work.

The following is the benchmark I've been optimizing for, though it also seems to somewhat improve the runtime of parts of the test suite (probably not so that we'd notice):

```python

from hypothesis import given, settings, strategies as st, seed
import traceback
import math


N_EXAMPLES = 10 ** 3


@seed(1)
@settings(
    database=None, timeout=-1, perform_health_check=False,
    max_examples=N_EXAMPLES, max_iterations=N_EXAMPLES,
)
@given(st.lists(st.floats(), min_size=10))
def test(fs):
    assert not math.isnan(sum(fs))


if __name__ == '__main__':
    try:
        test()
    except AssertionError:
        traceback.print_exc()
```

This runs in ~2.5s on my computer on master and in ~0.9 on this branch.

### CODE REVIEW

**Comment by DRMacIver** [comment]
Heh. Finally got around to figuring out why this was failing: A test relied on being able to trigger bad shrinking. When the shrinker improved it started failing because the shrinking it was triggering wasn't bad enough.

**Comment by DRMacIver** [comment]
Phew. OK @alexwlchan, this is finally actually working and ready for review. :-)

**Comment by alexwlchan** [comment]
I don’t have time to review this properly today, but will look at it properly in the next few days.

**Comment by DRMacIver** [comment]
👍 No problem. That wasn't a "Pls review now" nag, it's just that this PR has been sitting here in a broken state for ages so I wanted to make sure you knew it was now not broken.

**Comment by alexwlchan** [comment]
Also, sub 3-hour build time. :D

File: src/hypothesis/internal/compat.py [hunk 1]
```diff
@@ -392,7 +392,7 @@ def __add__(self, value):
         return compatbytes(bytearray.__add__(self, value))
 
     def __radd__(self, value):
-        return compatbytes(bytearray.__radd__(self, value))
+        return compatbytes(bytearray.__add__(value, self))
```
Comment (DRMacIver) [inline src/hypothesis/internal/compat.py]: This was actually just a bug that I ran into. It turns out bytearray.__radd__ doesn't exist, but this never actually got invoked in normal code (it got invoked in my code because I made a mistake and I fixed this to try to get a better error).

Properly I should add a test for this, but actually it might be better to just delete this method.

File: src/hypothesis/internal/conjecture/engine.py [hunk 2]
```diff
@@ -378,51 +378,48 @@ def _run(self):
...
+                k //= 2
 
             if change_counter != self.changed:
                 self.debug('Restarting')
                 continue
 
-            self.debug('Replacing blocks with simpler blocks')
+            self.debug('Bulk replacing blocks with simpler blocks')
+            i = 0
+            while i < len(self.last_data.blocks):
```
Comment (DRMacIver) [inline src/hypothesis/internal/conjecture/engine.py]: Yup.

Note that we don't mutate the list, we replace it with an entirely new one.

File: src/hypothesis/internal/conjecture/minimizer.py [hunk 3]
```diff
@@ -47,11 +47,15 @@ def __init__(self, initial, condition, random, cautious):
...
         self.changes = 0
+        self.seen = set()
 
     def incorporate(self, buffer):
         assert isinstance(buffer, hbytes)
         assert len(buffer) == self.size
         assert buffer <= self.current
+        if buffer in self.seen:
+            return False
+        self.seen.add(buffer)
```
Comment (DRMacIver) [inline src/hypothesis/internal/conjecture/minimizer.py]: Yup.

File: src/hypothesis/internal/conjecture/minimizer.py [hunk 4]
```diff
@@ -78,55 +86,95 @@ def _shrink_index(self, i, c):
...
+            while k > 0:
+                i = 0
+                while i + k <= len(self.current):
+                    self.incorporate(
+                        self.current[:i] +
+                        hbytes(f(b) for b in self.current[i:i + k]) +
+                        self.current[i + k:]
+                    )
+                    i += k
+                k //= 2
```
Comment (DRMacIver) [inline src/hypothesis/internal/conjecture/minimizer.py]: It's based on an algorithm called delta debugging. The idea is that we're trying to fixate where no improvement of a single value is permitted, but trying each value in turn is potentially slow if a lot of them are likely to be successful, so we try doing bulk replacement of values in large blocks first and then gradually shrink the size of the block down to 1.

Normally delta debugging is used with deleting, but it works just as well with rewriting.

File: src/hypothesis/internal/conjecture/minimizer.py [hunk 5]
```diff
@@ -78,55 +86,95 @@ def _shrink_index(self, i, c):
...
+                self.incorporate(rotated)
+
+    def shrink_indices(self):
+        for i in hrange(self.size):
+            t = self.current[i]
+            if t > 0:
+                ss = small_shrinks[self.current[i]]
+                for c in ss:
+                    if self._shrink_index(i, c):
+                        for c in hrange(self.current[i]):
```
Comment (DRMacIver) [inline src/hypothesis/internal/conjecture/minimizer.py]: No, because if we hit this point then self.current has changed, so self.current[i] < t.

File: src/hypothesis/internal/conjecture/minimizer.py [hunk 6]
```diff
@@ -78,55 +86,95 @@ def _shrink_index(self, i, c):
...
-                    ss = small_shrinks[self.current[i]]
-                    for c in ss:
-                        if self._shrink_index(i, c):
-                            for c in hrange(self.current[i]):
-                                if c in ss:
-                                    continue
-                                if self._shrink_index(i, c):
-                                    break
-                            break
+            self.rotate_suffixes()
```
Comment (DRMacIver) [inline src/hypothesis/internal/conjecture/minimizer.py]: I think the reasoning was that bulk operations should all happen before shrink\_indices, which is pretty expensive, but I don't remember if this reasoning was backed by benchmarking.

File: src/hypothesis/internal/conjecture/minimizer.py [hunk 7]
```diff
@@ -147,7 +195,7 @@ def run(self):
         result.add(b ^ (1 << i))
     result.discard(b)
     assert len(result) <= 10
-    small_shrinks.append(sorted(result))
+    small_shrinks.append(sorted(c for c in result if c < b))
```
Comment (DRMacIver) [inline src/hypothesis/internal/conjecture/minimizer.py]: Done

File: src/hypothesis/internal/compat.py (hunk 1 above)
Comment (alexwlchan) [inline src/hypothesis/internal/compat.py]: If it was never invoked in normal code, how did it get through the coverage checks? Why did we never spot the non-existence of `bytearray.__radd__`?

File: src/hypothesis/internal/conjecture/minimizer.py (hunk 4 above)
Comment (alexwlchan) [inline src/hypothesis/internal/conjecture/minimizer.py]: Oh yeah, this rings a bell now.

👍

**Comment by alexwlchan** [comment]
One question about previous testing, otherwise LGTM. 👍

File: src/hypothesis/internal/compat.py (hunk 1 above)
Comment (DRMacIver) [inline src/hypothesis/internal/compat.py]: compat.py is excluded from the coverage checks because there are large swathes of it that only get invoked on one version of Python or another and we only run coverage on Python 3.

We never spotted the non-existence of bytearray.__radd__ because this method was never invoked. We only ever add hbytes to hbytes (and don't expose hbytes to end users, thank the gods and/or past-DRMacIver), so this always goes through \_\_add\_\_ and \_\_radd\_\_ is never invoked. I only found this because I accidentally did str + hbytes, which goes via hbytes.\_\_radd\_\_ and triggers this problem.

Basically this method can only get invoked when we do something wrong.

**Comment by DRMacIver** [comment]
Nope! I'm all good (enthusiastic really) for it to be merged.
